File: 2024/python/day_01.py
"""Day 1"""


def puzzle_input(file):
    with open(file, "r+") as file:
        return zip(*[r.strip().split("   ") for r in file.readlines()])


def part1(x, y):
    """
    >>> x = [3, 4, 2, 1, 3, 3]
    >>> y = [4, 3, 5, 3, 9, 3]
    >>> part1(x, y)
    11
    """
    return sum(map(lambda i, j: abs(int(j) - int(i)), sorted(x), sorted(y)))


def part2(x, y):
    """
    >>> x = [3, 4, 2, 1, 3, 3]
    >>> y = [4, 3, 5, 3, 9, 3]
    >>> part2(x, y)
    31
    """
    return sum(map(lambda i: abs(int(i) * y.count(i)), x))


if __name__ == "__main__":
    x, y = puzzle_input("inputs/01.txt")
    print("Part 1:", part1(x, y))
    print("Part 2:", part2(x, y))

File: 2015/R/day_01.r
#' Day 01

puzzle_input <- function(file) {
  strsplit(readLines(file), "")[[1]]
}

part1 <- function(x) {
  sum(ifelse(x == "(", 1, -1))
}

part2 <- function(x) {
  which(cumsum(ifelse(x == "(", 1, -1)) < 0)[1]
}

x <- puzzle_input("inputs/01.txt")

cat("Part 1:", part1(x), "\n")
cat("Part 2:", part2(x), "\n")
